Overhaul of Resources

aaronpowell opened the pull request:
This is an update to the Resources and ResourceTemplates part of the MCP server which aims to align the data structures more with the structure that we see in the spec (using 2024-11-05 as the spec target).

* Fixed #123 by allowing a ListResourceTemplates handler to work in conjunction with ReadResourcesRequest
* Created new types for TextResourceContents and BlobResourceContents (inheriting ResourceContents) so that we can get better typing of the structures
* Updating tests and associated code

Rather than using `ResourceContents` as a combination of the properties across blob and text resources, I've turned this into an abstract class and have implementations of the types. This more closely aligns with the structure we see in the schema, making it easier to map our types to the source types.

stephentoub commented on src/ModelContextProtocol/Protocol/Types/Resources/BlobResourceContents.cs:
```diff
@@ -0,0 +1,16 @@
+public class BlobResourceContents : ResourceContents
+{
+    /// <summary>
+    /// The binary content of the resource.
```
```suggestion
    /// The base64-encoded string representing the binary data of the item.
```

stephentoub commented on src/ModelContextProtocol/Protocol/Types/Resources/Resource.cs:
```diff
@@ -3,7 +3,7 @@
 namespace ModelContextProtocol.Protocol.Types;
```
Why did you move all of these files?

aaronpowell replied:
Sorry, meant to revert that before pushing. I just wanted a way to find all the resources files in one quick view

stephentoub commented on src/ModelContextProtocol/Protocol/Types/Resources/ResourceContents.cs:
```diff
@@ -0,0 +1,120 @@
+    public string? MimeType { get; set; }
+}
+
+internal class ResourceContentsConverter : JsonConverter<ResourceContents>
```
It needs to be public to work with the source generator. I think in the past we've made them EditorBrowsableState.Never, and possibly nested them within the type they're converting. @eiriktsarpalis ?

aaronpowell replied:
What will the source generator be needed for in touching this?

stephentoub replied:
> What will the source generator be needed for in touching this?

The source generated by the source generator needs to be able to access the converter. If the serialization is being done in this assembly, internal is fine. But if another assembly using the source generator wants to serialize this type, the converter needs to be accessible to that generated source.

stephentoub commented on src/ModelContextProtocol/Protocol/Types/Resources/ResourceContents.cs:
```diff
@@ -0,0 +1,120 @@
+                Text = text
+            };
+        }
+        return null;
```
The resource must contain content? It can't just be a reference to some?

PederHP replied:
> The resource must contain content? It can't just be a reference to some?

In the spec all properties that have ResoureContents have "anyOf" TextResourceContents or BlobResourceContents. ResourceContents is only referenced in the spec as a base type for the other two. So this makes sense, I think, as a way to approximate the spec types functionally.

aaronpowell replied:
Looking at the TypeScript version of the spec, we can see it's a union type: https://github.com/modelcontextprotocol/specification/blob/main/schema/2024-11-05/schema.ts#L361-L363

If C# had unions... 🤪

stephentoub commented on src/ModelContextProtocol/AIContentExtensions.cs:
```diff
@@ -62,9 +58,12 @@ public static AIContent ToAIContent(this ResourceContents content)
+        {
+            BlobResourceContents blobResource => new DataContent(Convert.FromBase64String(blobResource.Blob), blobResource.MimeType ?? "application/octet-stream"),
+            TextResourceContents textResource => new TextContent(textResource.Text),
+            _ => throw new NotSupportedException($"Resource type '{content.GetType().Name}' is not supported.")
```
Should we add an internal ctor to ResourceContents such that this case isn't actually possible?

aaronpowell replied:
I've made it an abstract class, just to ensure that you can't create one which is "unknown", so that wouldn't be possible. But I'm unsure if raising errors like this is right from a server perspective.

stephentoub replied:
> I've made it an abstract class, just to ensure that you can't create one which is "unknown", so that wouldn't be possible.

Sure it is :smile:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABAzAAgIbADOALlHmMThgEw4BCehMAsAFADebO3OAlgHbEYUfngA29RjAAUAShzsAvm2Ws22HE3EwAJlVoARYbwBuunCElMFioA===

Not saying we need to, but it's definitely possible.

aaronpowell replied:
Well yes, you can add a constructor but you can't call it... can you?

stephentoub replied:
The derived types in the same assembly can. That's the point. If nothing outside of the assembly can derive from it, and all of the types inside the assembly are in the switch, then the default case can never be hit. With the type abstract and its default protected ctor, someone outside of the assembly can still derive from it, and then you need that default case to handle the unexoected input.

aaronpowell replied:
Ah cool, I hadn't considered that aspect

PederHP commented on src/ModelContextProtocol/Protocol/Types/Resources/TextResourceContents.cs:
```diff
@@ -0,0 +1,16 @@
+public class TextResourceContents : ResourceContents
+{
+    /// <summary>
+    /// The text content of the resource.
```
```suggestion
    /// The text of the item. This must only be set if the item can actually be represented as text (not binary data).
```
(from spec description)

stephentoub commented on src/ModelContextProtocol/Protocol/Types/BlobResourceContents.cs:
```diff
@@ -0,0 +1,16 @@
+    /// The base64-encoded string representing the binary data of the item.
+    /// </summary>
+    [JsonPropertyName("blob")]
+    public string Blob { get; set; } = default!;
```
We should either have this initialized to string.Empty or the property should be nullable.

stephentoub commented on src/ModelContextProtocol/Protocol/Types/Resource.cs:
```diff
@@ -32,6 +36,15 @@ public record Resource
+    /// This can be used by Hosts to display file sizes and estimate context window usage.
+    /// </summary>
+    [JsonPropertyName("size")]
+
```
```suggestion
```

stephentoub commented on src/ModelContextProtocol/Protocol/Types/ResourceContents.cs:
```diff
@@ -19,17 +26,106 @@ public class ResourceContents
+/// Converter for <see cref="ResourceContents"/>.
+/// </summary>
+[EditorBrowsable(EditorBrowsableState.Never)]
+public class ResourceContentsConverter : JsonConverter<ResourceContents>
```
Nit: as I'm doing in https://github.com/modelcontextprotocol/csharp-sdk/pull/145, let's make this a nested type of ResourceContents and just name it Converter.

stephentoub commented on src/ModelContextProtocol/Protocol/Types/ResourceContents.cs:
```diff
@@ -19,17 +26,106 @@ public class ResourceContents
+        string? blob = null;
+        string? text = null;
+
+        while (reader.Read())
+        {
+            if (reader.TokenType == JsonTokenType.EndObject)
+            {
+                break;
+            }
+
```
```suggestion
        while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
        {
```

stephentoub commented on src/ModelContextProtocol/Protocol/Types/ResourceContents.cs:
```diff
@@ -19,17 +26,106 @@ public class ResourceContents
+                break;
+            }
+
+            string? propertyName = reader.GetString();
+
+            if (propertyName == null)
+            {
+                continue;
+            }
```
```suggestion
            if (reader.TokenType != JsonTokenType.PropertyName)
            {
                continue;
            }

            string propertyName = reader.GetString();
```

stephentoub commented on src/ModelContextProtocol/Protocol/Types/ResourceContents.cs:
```diff
@@ -19,17 +26,106 @@ public class ResourceContents
+                Blob = blob
+            };
+        }
+        if (text is not null)
```
```suggestion

        if (text is not null)
```

stephentoub commented on src/ModelContextProtocol/Protocol/Types/ResourceContents.cs:
```diff
@@ -19,17 +26,106 @@ public class ResourceContents
+                Text = text
+            };
+        }
+        return null;
```
```suggestion

        return null;
```

stephentoub commented on src/ModelContextProtocol/Protocol/Types/ResourceContents.cs:
```diff
@@ -19,17 +26,106 @@ public class ResourceContents
+
+        writer.WriteStartObject();
+        writer.WriteString("uri", value.Uri);
+        writer.WriteString("mimeType", value.MimeType);
```
```suggestion
        writer.WriteString("mimeType", value.MimeType);
        Debug.Assert(value is BlobResourceContents or TextResourceContents);
```

stephentoub commented on tests/ModelContextProtocol.Tests/Server/McpServerResourceTests.cs:
```diff
@@ -0,0 +1,99 @@
+using ModelContextProtocol.Server;
+
+namespace ModelContextProtocol.Tests.Server;
+public class McpServerResourceTests
```
```suggestion

public class McpServerResourceTests
```

stephentoub reviewed (approved):
Once the remaining comments are addressed, LGTM. Thanks!

stephentoub commented on src/ModelContextProtocol/Protocol/Types/ResourceContents.cs:
```diff
@@ -52,20 +52,15 @@ public class ResourceContentsConverter : JsonConverter<ResourceContents>
                 continue;
             }
 
+            string propertyName = reader.GetString();
```
```suggestion
            string? propertyName = reader.GetString();
```

aaronpowell commented:
> Once the remaining comments are addressed, LGTM. Thanks!

Fixed that one and I think I got the merge correct from your change (and I fixed that at the same time)

jeffhandley commented:
Adding the `breaking-change` label retroactively during release notes revision.
